docs/cvl/hooks.md

```diff
@@ -20,77 +22,422 @@ Syntax
+```
+
+These different kinds of paths can be combined.  For example, the following
+hook will execute whenever the contract writes to the `balance` field of a
+struct in the `users` mapping of contract `C`:
+```cvl
+hook C.users[KEY address user].balance uint v (uint old_value) STORAGE { ... }
+```
+Inside the body of the hook, the variable `user` will refer to the address that
+was used as the key into the mapping `C.users`; the variable `v` will contain
+the value that is written, and the variable `old_value` will contain the value
+that was previously stored there.
+
+```{note}
+The only available access paths for `solc` versions 5.17 and older are `slot`
+and `offset` paths.
+```
+
+### Access path caveats
+
+```{todo}
+information on what kind of funny business happens if people do their own memory
+management, and about the disjointness of arrays and so forth
+
+How can you tell if an analysis failed and your hooks didn't apply?
+```
+
+(rawhooks)=
+### Hooking on all loads or stores
+
+Load and store hooks apply to reads and writes to specific storage locations.
+In some cases, it is useful to instrument every load or store, regardless of
+the location.
+
+The `ALL_SLOAD` and `ALL_SSTORE` opcode hooks are used for this purpose; they
+will be executed on every load and store instruction (in all contracts)
+respectively.  See {ref}`opcode-hooks` below for the general syntax of opcode
+hooks.
+
+The `ALL_SLOAD` opcode hook takes one input `uint` argument containing the slot
+number of the load instruction, and has one `uint` output containing the value
+that is loaded from the slot.  For example:
+```cvl
+hook ALL_SLOAD(uint slot) uint val { ... }
+```
+
+The `ALL_SSTORE` opcode hook takes two input `uint` arguments; the first is the
+slot number of the store instruction, and the second is the value being stored.
+For example:
+```cvl
+hook ALL_SSTORE(uint slot, uint val) { ... }
+```
+
+```{note}
+The storage splitting optimization must be disabled using the
+{ref}`-enableStorageSplitting` option in order to use the `ALL_SLOAD` or
+`ALL_SSTORE` hooks.
+```
+
+If a load instruction matches an `Sload` hook pattern and there is also an
+`ALL_SLOAD` hook, then both hooks will be executed; the `Sload` hook will apply
+first, and then the `ALL_SLOAD` hook.
+
+Similarly, if a store would trigger both an `Sstore` pattern and an `ALL_SSTORE`
+pattern, the `Sstore` hook would be executed, followed by the `ALL_SSTORE` hook.
+
+```{note}
+Just like the usual opcode hooks, the raw storage hooks are applied on all
+contracts.  This means that a storage access on _any_ contract will trigger the
+hook.  Therefore, in a rule that models calls to multiple contracts, if two
+contracts are accessing the same slot the same hook code will be called with
+the same slot number.
+```
+
+(opcode-hooks)=
+EVM opcode hooks
+----------------
+
+```{todo}
+Are `Create` hooks and `CREATE1`/`CREATE2` hooks different?  Should `Create`
```

Comment:
they are slightly different, let's skip documentation for `Create` to a later PR